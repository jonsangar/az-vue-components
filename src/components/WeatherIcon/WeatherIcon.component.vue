<template lang="pug">
.weatherIcon(
  v-if="code")
  img(
    v-if="getWeatherIcon"
    :width="size"
    :src="require(`@/assets/icons/weather/${getWeatherIcon}.svg`)"
  )

</template>

<script>
import getWeatherStatus from 'Units/utils/getWeatherStatus';

export default {

  props: {
    code: {
      type: Number,
      required: false
    },

    size: {
      type: String,
      default: '25px'
    },

    date: {
      type: String,
      default() {
        (new Date()).toString();
      }
    },

    sunriseToday: {
      type: String,
      required: false,
    },

    sunsetToday: {
      type: String,
      required: false,
    },

    sunriseTomorrow: {
      type: String,
      required: false,
    },

    sunsetTomorrow: {
      type: String,
      required: false,
    },
  },

  computed: {
    getWeatherIcon() {
    if(!this.code) return '';

    let isDay = true;

    const actualDate = new Date(this.date);

    if(this.sunriseToday && this.sunsetToday &&
      this.sunriseTomorrow && this.sunsetTomorrow){

      const sunriseTodayDate = new Date(this.sunriseToday);
      const sunsetTodayDate = new Date(this.sunsetToday);
      const sunriseTomorrowDate = new Date(this.sunriseTomorrow);
      const sunsetTomorrowDate = new Date(this.sunsetTomorrow);

      if(actualDate < sunriseTodayDate ){
        isDay = false;
      } else if(sunriseTodayDate < actualDate && actualDate < sunsetTodayDate) {
        isDay = true;
      } else if(sunsetTodayDate < actualDate && actualDate < sunriseTomorrowDate) {
        isDay = false;
      } else if(sunriseTomorrowDate < actualDate && actualDate < sunsetTomorrowDate) {
        isDay = true;
      } else {
        isDay = false;
      }
    }


    const weather = getWeatherStatus[this.code]

    if(weather === undefined) return null;

    return (weather.icon_night !== undefined && !isDay) ? weather.icon_night : weather.icon ;
  },
  }

}
</script>
