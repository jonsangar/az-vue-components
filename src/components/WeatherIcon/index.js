import WeatherIcon from './WeatherIcon.component.vue';

export default WeatherIcon;
