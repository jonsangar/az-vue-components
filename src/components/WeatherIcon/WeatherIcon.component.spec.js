import { shallowMount, createLocalVue } from '@vue/test-utils';
// Importamos librerías de estilos e iconos
import SvgIcon from 'vue-svgicon'
import WeatherIcon from './WeatherIcon.component.vue';

const localVue = createLocalVue()
localVue.use(SvgIcon, {
  tagName: 'az-icon',
  classPrefix: 'az',
})

describe('WeatherIcon.vue', () => {
  const build = () => {
    const wrapper = shallowMount(WeatherIcon, {
      localVue,
      propsData: {
        sunriseToday: '2021-12-01T07:11:51.000Z',
        sunsetToday: '2021-12-01T17:02:04.000Z',
        sunriseTomorrow: '2021-12-02T07:12:46.000Z',
        sunsetTomorrow: '2021-12-02T17:01:54.000Z',
      }
    })

    return {
      wrapper,
    }
  }


  it('Comprueba renderizado del componente WeatherIcon con snapshot', async() => {
    const { wrapper } = build();
    await wrapper.setProps({
      code: 800
    });

    expect(wrapper.html()).toMatchSnapshot()

  })

  it('Si no recibe un código no renderiza nada', () => {
    const { wrapper } = build();

    expect(wrapper.isVisible()).toBe(false);
  })

  it('Si recibe "code" renderiza el elemento', async () => {
    const { wrapper } = build();

    await wrapper.setProps({
      code: 800
    });

    expect(wrapper.classes()).toContain('weatherIcon')
  });

  it('Si renderiza elemento dibuja el icono', async () => {
    const { wrapper } = build();

    await wrapper.setProps({
      code: 800
    });
    // Se pinta <img> con el tamaño por defecto
    expect(wrapper.find('.weatherIcon').html()).toContain('<img width="25px" src=""></div>');

    await wrapper.setProps({
      size: '40px'
    });

    // Se pinta <img> con el tamaño pasado como prop
    expect(wrapper.find('.weatherIcon').html()).toContain('<img width="40px" src=""></div>');
  })

  it('Si recibe un código inválido no pinta la imagen', async () => {
    const { wrapper } = build();

    await wrapper.setProps({
      code: 10
    });

    // Se pinta <img> con el tamaño por defecto
    expect(wrapper.find('.weatherIcon').html()).not.toContain('<img width="25px" src=""></div>');
  });
});
