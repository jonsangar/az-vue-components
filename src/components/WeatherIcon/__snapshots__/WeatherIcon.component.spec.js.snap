// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`WeatherIcon.vue Comprueba renderizado del componente WeatherIcon con snapshot 1`] = `<div class="weatherIcon"><img width="25px" src=""></div>`;
